GVFS/GVFS.Common/Maintenance/MaintenanceJobManager.cs

```diff
@@ -0,0 +1,341 @@
+                    }
+                    else
+                    {
+                        this.maintenanceJobThread = new Thread(() => this.RunMaintenanceJobs());
+                        this.maintenanceJobThread.IsBackground = true;
+                        this.maintenanceJobThread.Start();
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+            catch (Exception e)
+            {
+                this.LogUnhandledExceptionAndExit(nameof(this.LaunchMaintenanceJobIfIdle), e);
+                return false;
+            }
+        }
+
+        public void TerminateJobs()
+        {
+            this.stopping = true;
+            lock (this.postFetchJobLock)
+            {
+                GitProcess process = this.maintenanceGitProcess;
+                if (process != null)
+                {
+                    if (process.TryKillRunningProcess())
+                    {
+                        this.tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            MaintenanceTelemetryKey + ": killed background Git process during " + nameof(this.TerminateJobs),
+                            metadata: null);
+                    }
+                    else
+                    {
+                        this.tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            MaintenanceTelemetryKey + ": failed to kill background Git process during " + nameof(this.TerminateJobs),
+                            metadata: null);
+                    }
+                }
+            }
+        }
+
+        public void LaunchPostFetchJob(List<string> packIndexes)
+        {
+            lock (this.postFetchJobLock)
+            {
+                if (this.postFetchJobThread?.IsAlive == true)
+                {
+                    this.tracer.RelatedWarning("Dropping post-fetch job since previous job is still running");
+                    return;
+                }
+
+                if (this.stopping)
+                {
+                    this.tracer.RelatedWarning("Dropping post-fetch job since attempting to unmount");
+                    return;
+                }
+
+                this.postFetchJobThread = new Thread(() => this.PostFetchJob(packIndexes));
+                this.postFetchJobThread.IsBackground = true;
+                this.postFetchJobThread.Start();
+            }
+        }
+
+        /// <summary>
+        /// This method is used for test purposes only.
+        /// </summary>
+        public void WaitForMaintenanceJobToFinish()
+        {
+            this.maintenanceJobThread?.Join();
+        }
+
+        private void RunMaintenanceJobs()
+        {
+            try
+            {
+                this.BackgroundPrefetch();

```

Comment:
I think this class already has too many responsibilities. It should not grow to handle even more.

It sounds like your primary goal is to have a shared implementation of "I need to run a thing in the background, and be able to shut it down in a reliable way". But that is a separate concern from the business logic of when/how to run cache maintenance, when/how to create a new midx, and when/how to clean up placeholder files. We should not mix all of those business logic concerns into the same class just because they need to share the same mechanism for doing background work.